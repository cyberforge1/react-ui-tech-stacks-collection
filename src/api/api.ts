// src/api/api.ts

import axios, { AxiosResponse } from 'axios';
import { Todo, TodoCreation, TodoUpdate, TodoListResponse, TodoResponse, ApiResponse } from '../types/types';

const API_BASE_URL = 'http://localhost:5001/api'; // Flask API base URL

// Get the main welcome message
export const getMainMessage = async (): Promise<string> => {
  try {
    const response: AxiosResponse<ApiResponse> = await axios.get(`${API_BASE_URL}/`);
    return response.data.message;
  } catch (error) {
    console.error('Error fetching Main message:', error);
    throw error;
  }
};

// Get Hello World message
export const getHelloWorld = async (): Promise<string> => {
  try {
    const response: AxiosResponse<ApiResponse> = await axios.get(`${API_BASE_URL}/helloworld/`);
    return response.data.message;
  } catch (error) {
    console.error('Error fetching Hello World message:', error);
    throw error;
  }
};

// Get all todos
export const getTodos = async (): Promise<Todo[]> => {
  try {
    const response: AxiosResponse<TodoListResponse> = await axios.get(`${API_BASE_URL}/todos/`);
    return response.data.todos;
  } catch (error) {
    console.error('Error fetching todos:', error);
    throw error;
  }
};

// Get a single todo by ID
export const getTodoById = async (id: number): Promise<Todo> => {
  try {
    const response: AxiosResponse<TodoResponse> = await axios.get(`${API_BASE_URL}/todos/${id}/`);
    return response.data.todo!;
  } catch (error) {
    console.error(`Error fetching todo with ID ${id}:`, error);
    throw error;
  }
};

// Create a new todo
export const createTodo = async (todo: TodoCreation): Promise<Todo> => {
  try {
    const response: AxiosResponse<TodoResponse> = await axios.post(`${API_BASE_URL}/todos/`, todo);
    return response.data.todo!;
  } catch (error) {
    console.error('Error creating todo:', error);
    throw error;
  }
};

// Update an existing todo
export const updateTodo = async (id: number, todoUpdate: TodoUpdate): Promise<Todo> => {
  try {
    const response: AxiosResponse<TodoResponse> = await axios.put(`${API_BASE_URL}/todos/${id}/`, todoUpdate);
    return response.data.todo!;
  } catch (error) {
    console.error(`Error updating todo with ID ${id}:`, error);
    throw error;
  }
};

// Delete a todo
export const deleteTodo = async (id: number): Promise<ApiResponse> => {
  try {
    const response: AxiosResponse<ApiResponse> = await axios.delete(`${API_BASE_URL}/todos/${id}/`);
    return response.data;
  } catch (error) {
    console.error(`Error deleting todo with ID ${id}:`, error);
    throw error;
  }
};
