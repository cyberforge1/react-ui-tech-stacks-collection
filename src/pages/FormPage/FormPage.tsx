// src/pages/FormPage/FormPage.tsx

import React, { useState, useEffect } from 'react';
import TodoForm from '../TodoForm/TodoForm';
import { getTodos, createTodo, updateTodo, deleteTodo } from '../../api/api';
import { Todo } from '../../types/types';

const FormPage: React.FC = () => {
  const [todos, setTodos] = useState<Todo[]>([]);
  const [isEditing, setIsEditing] = useState<number | null>(null);
  const [error, setError] = useState<string | null>(null);
  const [isLoading, setIsLoading] = useState<boolean>(false);

  useEffect(() => {
    loadTodos();
  }, []);

  const loadTodos = async () => {
    setIsLoading(true);
    try {
      const data = await getTodos();
      setTodos(data || []);
      setError(null);
    } catch {
      setError('Failed to load todos. Please try again later.');
      setTodos([]);
    } finally {
      setIsLoading(false);
    }
  };

  const handleAddTodo = async (title: string) => {
    try {
      await createTodo({ title });
      loadTodos();
    } catch {
      setError('Failed to add todo. Please try again.');
    }
  };

  const handleUpdateTodo = async (id: number, title: string) => {
    try {
      await updateTodo(id, { title });
      loadTodos();
      setIsEditing(null);
    } catch {
      setError('Failed to update todo. Please try again.');
    }
  };

  const handleDeleteTodo = async (id: number) => {
    try {
      await deleteTodo(id);
      loadTodos();
    } catch {
      setError('Failed to delete todo. Please try again.');
    }
  };

  return (
    <div data-testid="form-page">
      <h1>Todo Management</h1>
      {error && <p style={{ color: 'red' }}>{error}</p>}
      <TodoForm onSubmit={handleAddTodo} buttonText="Add Todo" />
      {isLoading ? (
        <p>Loading todos...</p>
      ) : (
        <ul>
          {todos.length > 0 ? (
            todos
              .slice() // Create a shallow copy to avoid mutating the original array
              .reverse() // Reverse the order of todos
              .map((todo) => (
                <li key={todo.id}>
                  {isEditing === todo.id ? (
                    <TodoForm
                      onSubmit={(title) => handleUpdateTodo(todo.id, title)}
                      initialTitle={todo.title}
                      buttonText="Update Todo"
                    />
                  ) : (
                    <>
                      <h3>{todo.title}</h3>
                      <button onClick={() => setIsEditing(todo.id)}>Edit</button>
                      <button onClick={() => handleDeleteTodo(todo.id)}>Delete</button>
                    </>
                  )}
                </li>
              ))
          ) : (
            <p>No todos available</p>
          )}
        </ul>
      )}
    </div>
  );
};

export default FormPage;
