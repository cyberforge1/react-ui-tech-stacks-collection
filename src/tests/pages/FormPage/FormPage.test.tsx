// src/tests/pages/FormPage/FormPage.test.tsx

import { render, screen, cleanup, fireEvent } from '@testing-library/react';
import { describe, it, expect, afterEach, vi } from 'vitest';
import FormPage from '../../../pages/FormPage/FormPage';

// Mock the API calls used in FormPage
vi.mock('../../../api/api', () => ({
  getTodos: vi.fn(),
  createTodo: vi.fn(),
  updateTodo: vi.fn(),
  deleteTodo: vi.fn(),
}));

import { getTodos, createTodo, updateTodo, deleteTodo } from '../../../api/api';

describe('FormPage Component', () => {
  afterEach(() => {
    cleanup();
    vi.clearAllMocks(); // Clear all mocks after each test
  });

  it('renders the heading "Todo Management"', () => {
    render(<FormPage />);
    const headingElement = screen.getByRole('heading', { name: /todo management/i });
    expect(headingElement).toBeTruthy();
  });

  it('displays "Loading todos..." when loading', () => {
    render(<FormPage />);
    const loadingText = screen.getByText(/loading todos.../i);
    expect(loadingText).toBeTruthy();
  });

  it('renders todos fetched from the API', async () => {
    (getTodos as ReturnType<typeof vi.fn>).mockResolvedValueOnce([{ id: 1, title: 'Test Todo' }]);

    render(<FormPage />);
    expect(await screen.findByText(/test todo/i)).toBeTruthy();
  });

  it('handles adding a new todo', async () => {
    (getTodos as ReturnType<typeof vi.fn>).mockResolvedValueOnce([]); // Initial fetch returns no todos
    (createTodo as ReturnType<typeof vi.fn>).mockResolvedValueOnce({ id: 2, title: 'New Todo' });
    (getTodos as ReturnType<typeof vi.fn>).mockResolvedValueOnce([{ id: 2, title: 'New Todo' }]); // After adding, fetch returns the new todo

    render(<FormPage />);
    const inputField = screen.getByPlaceholderText(/title/i);
    const addButton = screen.getByRole('button', { name: /add todo/i });

    fireEvent.change(inputField, { target: { value: 'New Todo' } });
    fireEvent.click(addButton);

    expect(await screen.findByText(/new todo/i)).toBeTruthy();
  });

  it('handles deleting a todo', async () => {
    (getTodos as ReturnType<typeof vi.fn>).mockResolvedValueOnce([{ id: 1, title: 'Test Todo' }]); // Initial fetch
    (deleteTodo as ReturnType<typeof vi.fn>).mockResolvedValueOnce({ message: 'Deleted' });
    (getTodos as ReturnType<typeof vi.fn>).mockResolvedValueOnce([]); // After delete, fetch returns no todos

    render(<FormPage />);
    const deleteButton = await screen.findByRole('button', { name: /delete/i });
    fireEvent.click(deleteButton);

    expect(deleteTodo).toHaveBeenCalledWith(1);
    expect(await screen.findByText(/no todos available/i)).toBeTruthy();
  });

  it('displays an error message when API call fails', async () => {
    (getTodos as ReturnType<typeof vi.fn>).mockImplementationOnce(() => Promise.reject(new Error('Failed to fetch todos')));

    render(<FormPage />);
    expect(await screen.findByText(/failed to load todos\. please try again later\./i)).toBeTruthy();
  });

  it('handles editing a todo', async () => {
    (getTodos as ReturnType<typeof vi.fn>).mockResolvedValueOnce([{ id: 1, title: 'Editable Todo' }]);
    (updateTodo as ReturnType<typeof vi.fn>).mockResolvedValueOnce({ id: 1, title: 'Updated Todo' });
    (getTodos as ReturnType<typeof vi.fn>).mockResolvedValueOnce([{ id: 1, title: 'Updated Todo' }]); // After update, fetch returns updated todo

    render(<FormPage />);
    const editButton = await screen.findByRole('button', { name: /edit/i });
    fireEvent.click(editButton);

    const inputField = screen.getByDisplayValue(/editable todo/i);
    const updateButton = screen.getByRole('button', { name: /update todo/i });

    fireEvent.change(inputField, { target: { value: 'Updated Todo' } });
    fireEvent.click(updateButton);

    expect(await screen.findByText(/updated todo/i)).toBeTruthy();
  });

  it('matches the snapshot', () => {
    const { container } = render(<FormPage />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
