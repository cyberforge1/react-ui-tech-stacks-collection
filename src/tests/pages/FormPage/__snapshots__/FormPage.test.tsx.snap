// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FormPage Component > matches the snapshot 1`] = `
<div
  data-testid="form-page"
>
  <h1>
    Todo Management
  </h1>
  <form
    data-testid="todo-form"
  >
    <input
      placeholder="Title"
      type="text"
      value=""
    />
    <button
      type="submit"
    >
      Add Todo
    </button>
  </form>
  <p>
    Loading todos...
  </p>
</div>
`;
